File: backend/src/main/java/ci/company/eduops/cashier/service/CashService.java
package ci.company.eduops.cashier.service;

import ci.company.eduops.cashier.domain.CashSession;
import ci.company.eduops.cashier.repository.CashSessionRepository;
import ci.company.eduops.common.exception.*;
import ci.company.eduops.common.tenant.TenantContext;
import ci.company.eduops.payment.repository.PaymentRepository;
import ci.company.eduops.payment.domain.PaymentStatus;
import ci.company.eduops.school.repository.SchoolRepository;
import ci.company.eduops.security.service.CurrentUser;
import ci.company.eduops.audit.service.AuditService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.*;

@Service
@Transactional
public class CashService {
    private final CashSessionRepository sessions;
    private final PaymentRepository payments;
    private final SchoolRepository schools;
    private final CurrentUser user;
    private final AuditService audit;
    public CashService(CashSessionRepository sessions, PaymentRepository payments, SchoolRepository schools, CurrentUser user, AuditService audit) {
        this.sessions=sessions; this.payments=payments; this.schools=schools; this.user=user; this.audit=audit;
    }
    public record Session(UUID id, String reference, String status, OffsetDateTime openedAt, OffsetDateTime closedAt,
        BigDecimal openingBalance, BigDecimal cashReceived, BigDecimal expectedBalance, BigDecimal actualBalance,
        BigDecimal difference, String notes) {}
    public record Movement(UUID id, String reference, String studentName, BigDecimal amount, String method, String date) {}
    private UUID school() {
        UUID id=TenantContext.getSchoolId();
        if(id==null) throw BusinessException.of(ErrorCode.SCHOOL_NOT_FOUND);
        return id;
    }
    private CashSession owned(CashSession s) {
        if (!school().equals(s.getSchool().getId()) || !user.requireId().equals(s.getCashierUserId()))
            throw BusinessException.of(ErrorCode.CASH_SESSION_NOT_FOUND);
        return s;
    }
    @Transactional(readOnly=true)
    public List<Session> list() {
        return sessions.findBySchoolIdAndCashierUserIdOrderByOpenedAtDesc(school(), user.requireId()).stream().map(this::response).toList();
    }
    @Transactional(readOnly=true)
    public List<Movement> movements(UUID id) {
        owned(sessions.findById(id).orElseThrow(() -> BusinessException.of(ErrorCode.CASH_SESSION_NOT_FOUND)));
        return payments.findByCashSessionIdAndStatus(id, PaymentStatus.VALIDATED).stream()
            .map(p -> new Movement(p.getId(), p.getPaymentReference(), p.getStudent().getFirstName()+" "+p.getStudent().getLastName(),
                p.getAmount(), p.getPaymentMethod().name(), p.getPaymentDate().toString())).toList();
    }
    public Session open(BigDecimal balance, String notes) {
        UUID schoolId=school(); UUID userId=user.requireId();
        if(sessions.findOpenForCashier(userId).isPresent()) throw BusinessException.of(ErrorCode.CASH_SESSION_ALREADY_OPEN);
        CashSession s=new CashSession();
        s.setSchool(schools.findById(schoolId).orElseThrow(() -> BusinessException.of(ErrorCode.SCHOOL_NOT_FOUND)));
        s.setCashierUserId(userId); s.setReference("CSH-"+UUID.randomUUID()); s.setOpeningBalance(balance); s.setNotes(notes);
        sessions.saveAndFlush(s);
        audit.logCreate("CashSession", s.getId(), s.getReference(), Map.of("openingBalance",balance));
        return response(s);
    }
    public Session close(UUID id, BigDecimal actual, BigDecimal reviewedExpected, String notes) {
        CashSession s=owned(sessions.lockById(id).orElseThrow(() -> BusinessException.of(ErrorCode.CASH_SESSION_NOT_FOUND)));
        if(!s.isOpen()) throw BusinessException.of(ErrorCode.CASH_SESSION_CLOSED);
        BigDecimal expected=s.getOpeningBalance().add(payments.sumCashForSession(id));
        if(expected.compareTo(reviewedExpected)!=0) throw BusinessException.of(ErrorCode.CONCURRENT_MODIFICATION,
            "Le solde a changé. Actualisez la caisse avant de confirmer la clôture.");
        if(actual.compareTo(expected)!=0 && (notes==null || notes.isBlank()))
            throw BusinessException.of(ErrorCode.VALIDATION_ERROR, "Expliquez l’écart de caisse.");
        s.close(expected,actual); s.setNotes(notes); sessions.saveAndFlush(s);
        audit.logValidate("CashSession",s.getId(),s.getReference(),notes);
        return response(s);
    }
    private Session response(CashSession s) {
        BigDecimal cash=s.isOpen()?payments.sumCashForSession(s.getId()):s.getExpectedBalance().subtract(s.getOpeningBalance());
        BigDecimal expected=s.getOpeningBalance().add(cash);
        return new Session(s.getId(),s.getReference(),s.getStatus().name(),s.getOpenedAt(),s.getClosedAt(),s.getOpeningBalance(),cash,
            expected,s.getActualBalance(),s.getActualBalance()==null?null:s.getActualBalance().subtract(expected),s.getNotes());
    }
}

File: frontend/src/app/core/services/cash.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment';
import { defer, of } from 'rxjs';
import { AuthService } from '@core/auth/auth.service';

export interface CashSession {
  id: string; reference: string; status: 'OPEN' | 'CLOSED' | 'RECONCILED'; openedAt: string; closedAt: string | null;
  openingBalance: number; cashReceived: number; expectedBalance: number; actualBalance: number | null;
  difference: number | null; notes: string;
}
export interface CashMovement { id: string; reference: string; studentName: string; amount: number; method: string; date: string; }
@Injectable({ providedIn: 'root' })
export class CashService {
  private readonly http = inject(HttpClient);
  private readonly auth = inject(AuthService);
  private readonly url = `${environment.apiBaseUrl}/cash/sessions`;
  private readonly demo = new Map<string, CashSession[]>();
  private key() { const u = this.auth.currentUser(); return `${u?.schoolId}:${u?.userId}`; }
  private items() { return this.demo.get(this.key()) ?? []; }
  private save(items: CashSession[]) { this.demo.set(this.key(), items); }
  list() { return environment.useMockData ? of(structuredClone(this.items())) : this.http.get<CashSession[]>(this.url); }
  movements(id: string) { return environment.useMockData ? of([] as CashMovement[]) : this.http.get<CashMovement[]>(`${this.url}/${id}/movements`); }
  open(openingBalance: number, notes: string) {
    if (!environment.useMockData) return this.http.post<CashSession>(this.url, { openingBalance, notes });
    return defer(() => {
      if (this.items().some(s => s.status === 'OPEN')) throw new Error('Une session est déjà ouverte.');
      const s: CashSession = { id: crypto.randomUUID(), reference: `CSH-DEMO-${this.items().length + 1}`, status: 'OPEN',
        openedAt: new Date().toISOString(), closedAt: null, openingBalance, cashReceived: 0, expectedBalance: openingBalance,
        actualBalance: null, difference: null, notes };
      this.save([s, ...this.items()]); return of(s);
    });
  }
  close(session: CashSession, actualBalance: number, notes: string) {
    if (!environment.useMockData) return this.http.post<CashSession>(`${this.url}/${session.id}/close`, { actualBalance, expectedBalance: session.expectedBalance, notes });
    return defer(() => {
      const current = this.items().find(i => i.id === session.id);
      if (!current || current.status !== 'OPEN') throw new Error('Cette session est clôturée.');
      const s: CashSession = { ...current, actualBalance, difference: actualBalance - current.expectedBalance,
        notes, status: 'CLOSED', closedAt: new Date().toISOString() };
      this.save(this.items().map(i => i.id === s.id ? s : i)); return of(s);
    });
  }
}
